src/update-labels-in-github.js

```diff
@@ -0,0 +1,166 @@
+  },
+  json: true
+};
+
+function readDeclaredLabelsFromConfig(config) {
+
+  const labels = [];
+
+  for (let category of config) {
+    const priorities = category['visual-priorities'];
+
+    if (!category.labels) {
+      continue;
+    }
+
+    for (let [labelName, labelMeta] of Object.entries(category.labels)) {
+      const labelPriority = labelMeta["visual-priority"];
+      const color = priorities[labelPriority].replace('#', '');
+
+      const fullName = category.name + ': ' + labelName;
+
+      labels.push({
+        name: fullName,
+        description: labelMeta.description,
+        color: color
+      });
+    }
+  }
+
+  return labels;
+}
+
+const declaredLabels = readDeclaredLabelsFromConfig(config);
+
+async function fetchReposForOrg(orgName) {
+  return await request.get(`https://api.github.com/orgs/${orgName}/repos?per_page=100`, opts);
+}
+
+async function fetchRepo(repoName) {
+  return await request.get(`https://api.github.com/repos/${repoName}`, opts);
+}
+
+async function addLabelToRepo(label, repoLabelsUrl) {
+  let reqOpts = Object.assign({}, opts, {
+    'json': label
+  });
+
+  return await request.post(repoLabelsUrl, reqOpts);
+}
+
+async function updateLabelInRepo(label, repoLabelsUrl) {
+  let reqOpts = Object.assign({}, opts, {
+    'json': label
+  });
+
+  const url = repoLabelsUrl + '/' + encodeURIComponent(label.name);
+
+  return await request.patch(url, reqOpts);
+}
+
+async function updateRepo(repo) {
+
+  const repoLabelsUrl = repo["labels_url"].replace(/{[^{}]+}/, '');
+  const remoteLabels = await request.get(repoLabelsUrl + "?per_page=100", opts);
+
+  const declaredLabelsToAdd = declaredLabels.filter(dl => !remoteLabels.some(rl => dl.name === rl.name));
+
+  let declaredLabelsToUpdate = [];
+
+  const declaredLabelsInRepo = declaredLabels.filter(dl => remoteLabels.some(rl => dl.name === rl.name));
+
+  for (let declaredLabel of declaredLabelsInRepo) {
+
+    const labelName = declaredLabel.name;
+
+    const remoteLabel = remoteLabels.find(l => l.name === labelName);
+
+    const needsUpdate = declaredLabel.description !== remoteLabel.description
+          || declaredLabel.color !== remoteLabel.color;
+
+    // if none of the fields we care about match
+    if (needsUpdate) {
+      declaredLabelsToUpdate.push(declaredLabel);
+    }
+
+  }
+
+  if (declaredLabelsToAdd.length === 0 && declaredLabelsToUpdate.length === 0) {
+    console.log(`Nothing to change for "${repo.name}"`);
+    return;
+  }
+
+  for (let declaredLabel of declaredLabelsToAdd) {
+    console.log(`adding label to "${repo.name}": "${declaredLabel.name}"`);
+    await addLabelToRepo(declaredLabel, repoLabelsUrl);
+  }
+
+  for (let declaredLabel of declaredLabelsToUpdate) {
+    console.log(`updating label to "${repo.name}": "${declaredLabel.name}"`);
+    await updateLabelInRepo(declaredLabel, repoLabelsUrl);
+  }
+
+}
+
+async function run() {
+
+  if (toUpdate.org) {
+    const repos = await fetchReposForOrg(toUpdate.org);
```

Comment:
Just wanted to check my understanding - Is the idea behind this that `request.get()` is asynchronous, but the result is required, so we force execution to wait for it to return before continuing?

Reply:
`async` and `await` are syntax around ES6 promises. It helps make asynchronous code look synchronous. An `async` function is a promise behind the scenes.

So, without async/await, this logic for the inside of this function would look roughly like the following (the function would also return a promise that would resolve successfully if the function ends successfully):
```js
if (toUpdate.org) {
  fetchReposForOrg(toUpdate.org)
      .then((repos) => {
        repos.forEach(r => updateRepo(r));
      });
}
```

Am I making sense there?

Reply:
I think this might contain most things you'd need to know about async/await. Let me know if you'd like to talk through it in any case. 

https://javascript.info/async-await

Reply:
Very clear, thanks very much 😄